LoongArch-ELF-ABI-EN.adoc

```diff
@@ -0,0 +1,206 @@
+Loongson Technology Corporation Limited
+V1.00
+
+:encoding: UTF-8
+:frame: top
+:grid: rows
+:valign: middle
+
+== Register Convention
+.Integer Register Convention
+[width="80%",options="header",cols="2m,2m,^5,^3"]
+|====================================================================
+|Name     |Alias        |Meaning             |Preserved across calls
+|$r0      |$zero        |Constant zero       |--
+|$r1      |$ra          |Return address      |No
+|$r2      |$tp          |Thread pointer      |--(Unallocatable)
+|$r3      |$sp          |Stack pointer       |Yes
+|$r4-$r11 |$a0-$a7      |Argument registers  |No
+|$r4-$r5  |$v0-$v1      |Return value        |No
+|$r12-$r20|$t0-$t8      |Temp registers      |No
+|$r21     |$x           |Reserved            |--(Unallocatable)
+|$r22     |$fp          |Frame pointer       |Yes
+|$r23-$r31|$s0-$s8      |Subroutine register variables   |Yes
+|====================================================================
+
+.Floating-point Register Convention
+[width="80%",options="header",cols="2m,2m,^5,^3"]
+|====================================================================
+|Name     |Alias        |Meaning            |Preserved across calls
+|$f0-$f7  |$fa0-$fa7    |Argument registers |No
+|$f0-$f1  |$fv0-$fv7    |Return value       |No
+|$f8-$f23 |$ft0-$ft15   |Temp registers     |No
+|$f24-$f31|$fs0-$fs7    |Subroutine register variables   |Yes
+|====================================================================
+
+
+== Type Size and Alignment
+
+.LP64 ABI
+[width="80%",options="header",cols="2m,^5,^3"]
+|====================================================================
+|Scalar type               |Size |Alignment
+|unsigned/signed char      |1    |1
+|unsigned/signed short     |2    |2
+|unsigned/signed int       |4    |4
+|unsigned/signed long      |8    |8
+|unsigned/signed long long |8    |8
+|pointer                   |8    |8
+|float                     |4    |4
+|double                    |8    |8
+|long double               |16   |16
+|====================================================================
+
+== ELF Object Files
+
+e_machine: LoongArch (258)
+
+e_flags: LP32 (0x1) LP64 (0x3)
+
+== Relocations
+.ELF Relocation types
+[width="100%",options="header",cols="^1,^1m,^10,10"]
+|====================================================================
+|Enum    |ELF reloc type          |Usage      |Detail
+|0       |R_LARCH_NONE            |           |
+
+|1	 |R_LARCH_32	          |Runtime address resolving|
+`*(int32_t *) PC = RtAddr + A`
+
+|2	 |R_LARCH_64	          |Runtime address resolving|
+`*(int64_t *) PC = RtAddr + A`
+
+|3       |R_LARCH_RELATIVE        |Runtime fixup for load-address|
+`*(void **) PC = B + A`
+
+|4       |R_LARCH_COPY            |Runtime memory copy in executable|
+`memcpy (PC, RtAddr, sizeof (sym))`
+
+|5       |R_LARCH_JUMP_SLOT       |Runtime PLT supporting|
+_implementation-defined_
+
+|6       |R_LARCH_TLS_DTPMOD32    |Runtime relocation for TLS-GD|
```

Comment:
Newer architectures should just use TLS descriptors. Drop traditional general-dynamic/local-dynamic relocations. You may implement them, but don't make them the default.

Reply:
Any docs about TLS descriptors?

Reply:
Search for x86-64 and aarch64 TLS descriptors.

https://maskray.me/blog/2021-02-14-all-about-thread-local-storage elaborated why the traditional model is awful.